examples/ensemble/plot_gradient_boosting_early_stopping.py

```diff
@@ -39,131 +43,135 @@
-    """
-    for i, rect in enumerate(rects):
-        plt.text(
-            rect.get_x() + rect.get_width() / 2.0,
-            1.05 * rect.get_height(),
-            "n_est=%d" % n_estimators[i],
-            ha="center",
-            va="bottom",
-        )
-
-
-autolabel(bar1, n_gb)
-autolabel(bar2, n_gbes)
-
-plt.ylim([0, 1.3])
-plt.legend(loc="best")
-plt.grid(True)
+# %%
+# Error Calculation
+# -----------------
+# The code calculates the :func:`~sklearn.metrics.mean_squared_error` for both
+# training and validation datasets for the models trained in the previous
+# section. It computes the errors for each boosting iteration. The purpose is
+# to assess the performance and convergence of the models.
+
+train_errors_without = []
+val_errors_without = []
+
+train_errors_with = []
+val_errors_with = []
+
+for i, (train_pred, val_pred) in enumerate(
+    zip(
+        gbm_no_early_stopping.staged_predict(X_train),
+        gbm_no_early_stopping.staged_predict(X_val),
+    )
+):
+    train_errors_without.append(mean_squared_error(y_train, train_pred))
+    val_errors_without.append(mean_squared_error(y_val, val_pred))
+
+for i, (train_pred, val_pred) in enumerate(
+    zip(
+        gbm_with_early_stopping.staged_predict(X_train),
+        gbm_with_early_stopping.staged_predict(X_val),
+    )
+):
+    train_errors_with.append(mean_squared_error(y_train, train_pred))
+    val_errors_with.append(mean_squared_error(y_val, val_pred))
 
-plt.xlabel("Datasets")
-plt.ylabel("Test score")
+# %%
+# Visualize Comparision
+# ---------------------
+# It includes three subplots:
+# 1. Plotting training errors of both models over boosting iterations.
+# 2. Plotting validation errors of both models over boosting iterations.
+# 3. Creating a bar chart to compare the training times and the estimator used
+# of the models with and without early stopping.
+
+plt.figure(figsize=(15, 6))
+
+plt.subplot(1, 3, 1)
+plt.plot(train_errors_without, label="Training Error (No Early Stopping)")
+plt.plot(train_errors_with, label="Training Error (With Early Stopping)")
+plt.xlabel("Boosting Iterations")
+plt.ylabel("MSE (Training)")
+plt.legend()
+
+plt.subplot(1, 3, 2)
+plt.plot(val_errors_without, label="Validation Error (No Early Stopping)")
+plt.plot(val_errors_with, label="Validation Error (With Early Stopping)")
+plt.xlabel("Boosting Iterations")
+plt.ylabel("MSE (Validation)")
+plt.legend()
+
+plt.subplot(1, 3, 3)
+training_times = [training_time_without, training_time_with]
+labels = ["No Early Stopping", "With Early Stopping"]
+bars = plt.bar(labels, training_times)
+plt.ylabel("Training Time (s)")
+
+for bar, n_estimators in zip(bars, [estimators_without, estimators_with]):
+    height = bar.get_height()
+    plt.text(
+        bar.get_x() + bar.get_width() / 2,
+        height + 0.01,
+        f"Estimators: {n_estimators}",
+        ha="center",
+        va="bottom",
+    )
 
+plt.tight_layout()
 plt.show()
```

Comment:
fig, axes = plt.subplots(ncols=3, figsize=(12, 4))

axes[0].plot(train_errors_without, label="gbm_full")
axes[0].plot(train_errors_with, label="gbm_early_stopping")
axes[0].set_xlabel("Boosting Iterations")
axes[0].set_ylabel("MSE (Training)")
axes[0].set_yscale("log")
axes[0].legend()
axes[0].set_title("Training Error")

axes[1].plot(val_errors_without, label="gbm_full")
axes[1].plot(val_errors_with, label="gbm_early_stopping")
axes[1].set_xlabel("Boosting Iterations")
axes[1].set_ylabel("MSE (Validation)")
axes[1].set_yscale("log")
axes[1].legend()
axes[1].set_title("Validation Error")

training_times = [training_time_without, training_time_with]
labels = ["gbm_full", "gbm_early_stopping"]
bars = axes[2].bar(labels, training_times)
axes[2].set_ylabel("Training Time (s)")

for bar, n_estimators in zip(bars, [estimators_without, estimators_with]):
    height = bar.get_height()
    axes[2].text(
        bar.get_x() + bar.get_width() / 2,
        height + 0.001,
        f"Estimators: {n_estimators}",
        ha="center",
        va="bottom",
    )

plt.tight_layout()
plt.show()